llvm/lib/Target/DirectX/DXILPostOptimizationValidation.cpp

```diff
@@ -0,0 +1,119 @@
+//===- DXILPostOptimizationValidation.cpp - Opt DXIL validation ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "DXILPostOptimizationValidation.h"
+#include "DXILConstants.h"
+#include "DXILIntrinsicExpansion.h"
+#include "DXILOpBuilder.h"
+#include "DXILShaderFlags.h"
+#include "DirectX.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/DXILMetadataAnalysis.h"
+#include "llvm/Analysis/DXILResource.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsDirectX.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/ErrorHandling.h"
+#include <cstdio>
+
+#define DEBUG_TYPE "dxil-post-optimization-validation"
+
+using namespace llvm;
+using namespace llvm::dxil;
+
+namespace {
+class DXILValidator {
+  Module &M;
+  DXILResourceMap &DRM;
+
+public:
+  DXILValidator(Module &M, DXILResourceMap &DRM) : M(M), DRM(DRM) {}
+
+  void validate() {
+    for (const auto &UAV : DRM.uavs()) {
+      if (UAV.CounterDirection != ResourceCounterDirection::Invalid)
+        continue;
+
+      CallInst *ResourceHandle = nullptr;
+      for (CallInst *MaybeHandle : DRM.calls()) {
+        if (*DRM.find(MaybeHandle) == UAV) {
+          ResourceHandle = MaybeHandle;
+          break;
+        }
+      }
+
+      StringRef Message =
+          "RWStructuredBuffers may increment or decrement their "
+          "counters, but not both.";
+      for (const auto &U : ResourceHandle->users()) {
+        const CallInst *CI = dyn_cast<CallInst>(U);
+        if (!CI && CI->getIntrinsicID() != Intrinsic::dx_resource_updatecounter)
+          continue;
+
+        M.getContext().diagnose(DiagnosticInfoGenericWithLoc(
+            Message, *CI->getFunction(), CI->getDebugLoc()));
+      }
+    }
+  }
+};
+} // namespace
+
+PreservedAnalyses
+DXILPostOptimizationValidation::run(Module &M, ModuleAnalysisManager &MAM) {
+  DXILResourceMap &DRM = MAM.getResult<DXILResourceAnalysis>(M);
+
+  DXILValidator(M, DRM).validate();
```

Comment:
The `DXILResourceMap` could expose `containsInvalidDirection()` flag that would be set during the `populateCounterDirections` analysis. If we knew that all the directions are correct, we can avoid iterating over all of the UAVs, and then this would be:

```suggestion
  if (DRM.containsInvalidDirection())
    reportInvalidDirection(M, DRM);
```

I am using this pattern in the `ResourceBindingAnalysis` PR [here](https://github.com/llvm/llvm-project/pull/137258/files#diff-d664aae7a8a1ad36eaf2012b19372305aeb1e2a58c46c297665b30eaee4d272fR664) for overlapping and implicit bindings. Reporting of overlapping bindings can be then added to this pass the same way:

```
if (BindingAnalysisResult.containsOverlappingBinding())
  reportOverlappingBinding();
```